File: legacy-python/rules.py
"""Deterministic Dutch motor-claim rules engine.

Legal anchors (advisory — implementation MUST be reviewed by legal counsel
before production use):

- WAM (Wet aansprakelijkheidsverzekering motorrijtuigen) art. 3: mandatory
  third-party liability insurance for motor vehicles in NL.
- WAM art. 22: statutory minimum coverage (currently EUR 6.45M property /
  EUR 1.3B personal injury per event).
- BW art. 3:310: 5-year limitation from knowledge of damage & liable party.
- BW art. 6:98 / 6:162: causation & unlawful act (grondslag for liability).
- WVW art. 185: strict liability of motor vehicle owner toward
  non-motorised road users (pedestrians / cyclists).
- Notification: WAM does not fix a hard deadline, but standard NL policies
  require notification "zo spoedig mogelijk" — 30 days is used here as a
  soft outer bound flag.

The rules engine NEVER auto-denies. Denials always route to a human.
"""
import os
import re
from datetime import date, datetime
from typing import Any

VIN_RE = re.compile(r"^[A-HJ-NPR-Z0-9]{17}$")
NL_PLATE_RE = re.compile(r"^[A-Z0-9]{5,8}$")  # normalised (dashes stripped)


def _cap() -> float:
    return float(os.environ.get("AUTO_APPROVE_CAP_EUR", 2500))


def _conf_min() -> float:
    return float(os.environ.get("EXTRACTION_CONFIDENCE_MIN", 0.85))


def _parse_date(s: str | None) -> date | None:
    if not s:
        return None
    try:
        return datetime.strptime(s[:10], "%Y-%m-%d").date()
    except (ValueError, TypeError):
        return None


def evaluate(claim: dict, documents: list[dict], fraud_score: float) -> dict:
    """Run the deterministic checks. Return {outcome, reasons, trace}.

    outcome: auto_approved | assisted | manual
    reasons: list of {code, ok, message, legal_ref?}
    trace: dict — the raw values checked (audit trail)
    """
    reasons: list[dict] = []
    trace: dict[str, Any] = {}

    # --- 1. Basic FNOL completeness -------------------------------------------
    required = ["policyholder_name", "policy_number", "license_plate",
                "loss_date", "loss_location", "description"]
    missing = [k for k in required if not claim.get(k)]
    trace["missing_fnol_fields"] = missing
    reasons.append({
        "code": "FNOL_COMPLETE",
        "ok": not missing,
        "message": "All required FNOL fields present." if not missing
                   else f"Missing FNOL fields: {', '.join(missing)}.",
    })

    # --- 2. Plate / VIN format -------------------------------------------------
    plate = (claim.get("license_plate") or "").upper().replace("-", "")
    vin = (claim.get("vin") or "").upper()
    plate_ok = bool(NL_PLATE_RE.match(plate)) if plate else False
    vin_ok = bool(VIN_RE.match(vin)) if vin else True  # VIN optional at FNOL
    trace["plate"] = plate
    trace["vin"] = vin
    reasons.append({
        "code": "VEHICLE_ID_FORMAT",
        "ok": plate_ok and vin_ok,
        "message": (
            "Plate/VIN format valid." if plate_ok and vin_ok
            else "Plate or VIN has unexpected format. Adjuster review."
        ),
    })

    # --- 3. Loss date + notification window ----------------------------------
    loss = _parse_date(claim.get("loss_date"))
    today = date.today()
    days_since = (today - loss).days if loss else None
    trace["days_since_loss"] = days_since
    date_ok = loss is not None and loss <= today
    reasons.append({
        "code": "LOSS_DATE_VALID",
        "ok": date_ok,
        "message": "Loss date parseable and not in the future." if date_ok
                   else "Loss date missing or in the future.",
        "legal_ref": "BW 3:310 (limitation 5y)",
    })

    within_notice = days_since is not None and 0 <= days_since <= 30
    reasons.append({
        "code": "NOTIFICATION_TIMELY",
        "ok": within_notice,
        "message": (f"Reported {days_since} days after loss." if days_since is not None
                    else "Cannot determine notification delay."),
        "legal_ref": "Policy: notification zo spoedig mogelijk",
    })

    # BW 3:310 5-year absolute cut-off for limitation
    limitation_ok = days_since is not None and days_since <= 5 * 365
    reasons.append({
        "code": "WITHIN_LIMITATION",
        "ok": limitation_ok,
        "message": "Within 5-year limitation window." if limitation_ok
                   else "Outside 5-year limitation (BW 3:310). Manual review.",
        "legal_ref": "BW 3:310",
    })

    # --- 4. Injuries / third-party liability ---------------------------------
    injuries = bool(claim.get("injuries"))
    third_party = bool(claim.get("third_party_involved"))
    trace["injuries"] = injuries
    trace["third_party_involved"] = third_party
    reasons.append({
        "code": "NO_PERSONAL_INJURY",
        "ok": not injuries,
        "message": "No personal injury reported." if not injuries
                   else "Personal injury reported — WVW 185 / BW 6:162 exposure. Manual review.",
        "legal_ref": "WVW 185, BW 6:162",
    })
    reasons.append({
        "code": "NO_THIRD_PARTY_DISPUTE",
        "ok": not third_party,
        "message": "First-party only." if not third_party
                   else "Third-party involved — WAM liability path. Assisted review.",
        "legal_ref": "WAM art. 3",
    })

    # --- 5. Amount envelope ---------------------------------------------------
    amt = claim.get("estimated_amount_eur")
    cap = _cap()
    trace["estimated_amount_eur"] = amt
    trace["auto_approve_cap_eur"] = cap
    amt_ok = amt is not None and 0 < amt <= cap
    reasons.append({
        "code": "AMOUNT_WITHIN_CAP",
        "ok": amt_ok,
        "message": (
            f"Estimated EUR {amt:.2f} within auto-approve cap EUR {cap:.2f}."
            if amt_ok else
            f"Estimated amount EUR {amt} exceeds cap EUR {cap:.2f} or is missing."
        ),
    })

    # --- 6. Extraction confidence --------------------------------------------
    conf = claim.get("extraction_confidence")
    conf_min = _conf_min()
    trace["extraction_confidence"] = conf
    trace["extraction_confidence_min"] = conf_min
    conf_ok = conf is not None and conf >= conf_min
    reasons.append({
        "code": "EXTRACTION_CONFIDENT",
        "ok": conf_ok,
        "message": (
            f"Extraction confidence {conf:.2f} ≥ threshold {conf_min:.2f}."
            if conf_ok else
            f"Extraction confidence {conf} below threshold {conf_min:.2f}."
        ),
    })

    # --- 7. Fraud signals -----------------------------------------------------
    trace["fraud_score"] = fraud_score
    fraud_ok = fraud_score < 0.3
    reasons.append({
        "code": "FRAUD_SIGNALS_LOW",
        "ok": fraud_ok,
        "message": (f"Fraud score {fraud_score:.2f} below 0.30 threshold."
                    if fraud_ok else
                    f"Fraud signals present (score {fraud_score:.2f}). Manual review."),
    })

    # --- 8. Total loss --------------------------------------------------------
    dmg_cats = claim.get("damage_categories") or "[]"
    total_loss_flagged = "total_loss" in (dmg_cats if isinstance(dmg_cats, str) else "")
    reasons.append({
        "code": "NOT_TOTAL_LOSS",
        "ok": not total_loss_flagged,
        "message": "Not classified as total loss." if not total_loss_flagged
                   else "Total loss classified — requires salvage/valuation review.",
    })

    # --- 9. Supporting evidence present --------------------------------------
    has_photo = any(d.get("doc_type") == "photo" for d in documents)
    has_estimate = any(d.get("doc_type") == "repair_estimate" for d in documents)
    evidence_ok = has_photo and has_estimate
    reasons.append({
        "code": "EVIDENCE_MINIMUM",
        "ok": evidence_ok,
        "message": ("At least one photo and one repair estimate on file."
                    if evidence_ok else
                    "Missing photo and/or repair estimate. Adjuster to request."),
    })

    # --- Outcome tiering ------------------------------------------------------
    all_pass = all(r["ok"] for r in reasons)
    hard_blockers = {"NO_PERSONAL_INJURY", "WITHIN_LIMITATION",
                     "FRAUD_SIGNALS_LOW", "LOSS_DATE_VALID",
                     "NOT_TOTAL_LOSS"}
    hard_fail = any(not r["ok"] and r["code"] in hard_blockers for r in reasons)

    if all_pass:
        outcome = "auto_approved"
    elif hard_fail:
        outcome = "manual"
    else:
        outcome = "assisted"

    return {"outcome": outcome, "reasons": reasons, "trace": trace}


def compute_fraud(claim: dict, documents: list[dict],
                  duplicate_photo_hits: int,
                  photo_signals: list[dict] | None = None) -> dict:
    """Additive fraud scoring. Returns {score: 0..1, signals: [...]}.

    Signals include type-of-check + severity so the UI can flag them.
    Score > 0.3 blocks auto-approval via the AMOUNT/FRAUD rule gate.

    ponytail: additive heuristic. Upgrade path — swap for a trained classifier
    once leakage/false-approval telemetry justifies it.
    """
    signals: list[dict] = []
    score = 0.0

    # 1. Recycled photo across prior claims — strongest single signal.
    if duplicate_photo_hits > 0:
        score += 0.6
        signals.append({
            "code": "PHOTO_RECYCLED",
            "severity": "high",
            "message": f"Photo perceptual-hash matched {duplicate_photo_hits} document(s) in prior claim(s).",
        })

    # 2. Delayed reporting (NL practice: sooner is normal, months late is suspicious).
    loss = _parse_date(claim.get("loss_date"))
    if loss:
        days = (date.today() - loss).days
        if days > 180:
            score += 0.2
            signals.append({
                "code": "REPORT_VERY_LATE",
                "severity": "medium",
                "message": f"Reported {days} days after loss (>180d).",
            })
        elif days > 60:
            score += 0.1
            signals.append({
                "code": "REPORT_LATE",
                "severity": "low",
                "message": f"Reported {days} days after loss (>60d).",
            })

    # 3. No supporting evidence at all.
    if not documents:
        score += 0.15
        signals.append({
            "code": "NO_EVIDENCE",
            "severity": "medium",
            "message": "No documents or photos submitted.",
        })

    # 4. Layered risk: injuries + third party + high estimate.
    if claim.get("injuries") and claim.get("third_party_involved"):
        score += 0.1
        signals.append({
            "code": "INJURY_THIRD_PARTY_LAYERED",
            "severity": "medium",
            "message": "Personal injury AND third-party liability reported — layered exposure.",
        })

    # 5. Photo-specific signals surfaced by app.py (EXIF mismatch, no EXIF).
    for ps in (photo_signals or []):
        sev = ps.get("severity", "low")
        signals.append(ps)
        score += {"high": 0.4, "medium": 0.2, "low": 0.05}.get(sev, 0.05)

    # 6. Very high estimate with low extraction confidence.
    amt = claim.get("estimated_amount_eur") or 0
    conf = claim.get("extraction_confidence") or 0
    if amt > 5000 and conf and conf < 0.6:
        score += 0.1
        signals.append({
            "code": "AMOUNT_HIGH_CONF_LOW",
            "severity": "medium",
            "message": f"Estimate €{amt:.0f} with low extraction confidence ({conf:.2f}).",
        })

    return {"score": min(score, 1.0), "signals": signals}


def compute_fraud_score(claim, documents, duplicate_photo_hits):
    """Back-compat shim. Prefer compute_fraud()."""
    return compute_fraud(claim, documents, duplicate_photo_hits)["score"]

File: src/JbAutoAi/Rules.cs
using System.Text.RegularExpressions;

namespace JbAutoAi;

/// Deterministic Dutch motor-claim rules engine.
///
/// Legal anchors are advisory and MUST be reviewed by counsel before production.
/// Each rule carries the citation key of a legal_refs row so the workspace can
/// show the source article behind the gate (see Legal.cs).
///
///  - WAM art. 2/3   mandatory motor liability insurance and its scope
///  - WAM art. 22    statutory minimum insured sums
///  - BW  art. 3:310 limitation: 5 years from knowledge, 20 years absolute
///  - BW  art. 6:98  causation / attribution of damage
///  - BW  art. 6:162 unlawful act
///  - BW  art. 7:941 duty to notify the insurer as soon as reasonably possible
///  - WVW art. 185   strict liability of the motorist toward non-motorised users
///
/// The engine NEVER auto-denies. Denials always route to a human.
public static partial class Rules
{
    [GeneratedRegex("^[A-HJ-NPR-Z0-9]{17}$")] private static partial Regex VinRe();
    [GeneratedRegex("^[A-Z0-9]{5,8}$")] private static partial Regex PlateRe();   // normalised, dashes stripped

    public record Reason(string Code, bool Ok, string Message, string? LegalRef = null);

    public record RulesResult(string Outcome, List<Reason> Reasons, Dictionary<string, object?> Trace);

    /// Weight is what this signal contributed to the composite score. Carrying it
    /// on the signal is what makes FR-7.4's "transparent signal breakdown" possible
    /// — the workspace can show the arithmetic instead of a bare number.
    public record Signal(string Code, string Severity, string Message, double Weight = 0);

    public static double SeverityWeight(string severity) =>
        severity switch { "high" => 0.4, "medium" => 0.2, _ => 0.05 };

    public record FraudResult(double Score, List<Signal> Signals);

    static readonly HashSet<string> HardBlockers =
    [
        "NO_PERSONAL_INJURY", "WITHIN_LIMITATION", "FRAUD_SIGNALS_LOW", "LOSS_DATE_VALID",
        "NOT_TOTAL_LOSS", "CITATIONS_RESOLVABLE",
    ];

    public static double Cap() => EnvDouble("AUTO_APPROVE_CAP_EUR", 2500);
    public static double ConfidenceMin() => EnvDouble("EXTRACTION_CONFIDENCE_MIN", 0.85);

    static double EnvDouble(string key, double fallback) =>
        double.TryParse(Environment.GetEnvironmentVariable(key),
                        System.Globalization.NumberStyles.Float,
                        System.Globalization.CultureInfo.InvariantCulture, out var v) ? v : fallback;

    public static RulesResult Evaluate(Claim claim, IReadOnlyList<Document> documents, double fraudScore,
                                       double citationIntegrity = 1.0, bool aiServicesHealthy = true)
    {
        var reasons = new List<Reason>();
        var trace = new Dictionary<string, object?>();

        // 1. FNOL completeness.
        var missing = new List<string>();
        if (string.IsNullOrWhiteSpace(claim.PolicyholderName)) missing.Add("policyholder_name");
        if (string.IsNullOrWhiteSpace(claim.PolicyNumber)) missing.Add("policy_number");
        if (string.IsNullOrWhiteSpace(claim.LicensePlate)) missing.Add("license_plate");
        if (claim.LossDate is null) missing.Add("loss_date");
        if (string.IsNullOrWhiteSpace(claim.LossLocation)) missing.Add("loss_location");
        if (string.IsNullOrWhiteSpace(claim.Description)) missing.Add("description");
        trace["missing_fnol_fields"] = missing;
        reasons.Add(new("FNOL_COMPLETE", missing.Count == 0,
            missing.Count == 0
                ? "All required FNOL fields present."
                : $"Missing FNOL fields: {string.Join(", ", missing)}."));

        // 2. Plate / VIN format.
        var plate = Db.NormalisePlate(claim.LicensePlate);
        var vin = (claim.Vin ?? "").ToUpperInvariant();
        var plateOk = plate.Length > 0 && PlateRe().IsMatch(plate);
        var vinOk = vin.Length == 0 || VinRe().IsMatch(vin);   // VIN optional at FNOL
        trace["plate"] = plate;
        trace["vin"] = vin;
        reasons.Add(new("VEHICLE_ID_FORMAT", plateOk && vinOk,
            plateOk && vinOk
                ? "Plate/VIN format valid."
                : "Plate or VIN has unexpected format. Adjuster review."));

        // 3. Loss date, notification window, limitation.
        var loss = claim.LossDate;
        var today = DateOnly.FromDateTime(DateTime.UtcNow);
        int? daysSince = loss is null ? null : today.DayNumber - loss.Value.DayNumber;
        trace["days_since_loss"] = daysSince;

        var dateOk = loss is not null && loss <= today;
        reasons.Add(new("LOSS_DATE_VALID", dateOk,
            dateOk ? "Loss date parseable and not in the future." : "Loss date missing or in the future.",
            "BW 3:310"));

        var withinNotice = daysSince is >= 0 and <= 30;
        reasons.Add(new("NOTIFICATION_TIMELY", withinNotice,
            daysSince is not null
                ? $"Reported {daysSince} days after loss."
                : "Cannot determine notification delay.",
            "BW 7:941"));

        var limitationOk = daysSince is not null && daysSince <= 5 * 365;
        reasons.Add(new("WITHIN_LIMITATION", limitationOk,
            limitationOk
                ? "Within 5-year limitation window."
                : "Outside 5-year limitation (BW 3:310). Manual review.",
            "BW 3:310"));

        // 4. Injuries / third-party liability.
        trace["injuries"] = claim.Injuries;
        trace["third_party_involved"] = claim.ThirdPartyInvolved;
        reasons.Add(new("NO_PERSONAL_INJURY", !claim.Injuries,
            !claim.Injuries
                ? "No personal injury reported."
                : "Personal injury reported — WVW 185 / BW 6:162 exposure. Manual review.",
            "WVW 185"));
        reasons.Add(new("NO_THIRD_PARTY_DISPUTE", !claim.ThirdPartyInvolved,
            !claim.ThirdPartyInvolved
                ? "First-party only."
                : "Third-party involved — WAM liability path. Assisted review.",
            "WAM 3"));

        // 5. Amount envelope.
        var amount = claim.EstimatedAmountEur;
        var cap = Cap();
        trace["estimated_amount_eur"] = amount;
        trace["auto_approve_cap_eur"] = cap;
        var amountOk = amount is > 0 && amount <= cap;
        reasons.Add(new("AMOUNT_WITHIN_CAP", amountOk,
            amountOk
                ? $"Estimated EUR {amount:F2} within auto-approve cap EUR {cap:F2}."
                : $"Estimated amount EUR {(amount?.ToString("F2") ?? "—")} exceeds cap EUR {cap:F2} or is missing."));

        // 6. Extraction confidence.
        var conf = claim.ExtractionConfidence;
        var confMin = ConfidenceMin();
        trace["extraction_confidence"] = conf;
        trace["extraction_confidence_min"] = confMin;
        var confOk = conf is not null && conf >= confMin;
        reasons.Add(new("EXTRACTION_CONFIDENT", confOk,
            confOk
                ? $"Extraction confidence {conf:F2} ≥ threshold {confMin:F2}."
                : $"Extraction confidence {(conf?.ToString("F2") ?? "—")} below threshold {confMin:F2}."));

        // 7. Fraud signals.
        trace["fraud_score"] = fraudScore;
        var fraudOk = fraudScore < 0.3;
        reasons.Add(new("FRAUD_SIGNALS_LOW", fraudOk,
            fraudOk
                ? $"Fraud score {fraudScore:F2} below 0.30 threshold."
                : $"Fraud signals present (score {fraudScore:F2}). Manual review."));

        // 8. Total loss.
        var totalLoss = (claim.DamageCategories ?? "").Contains("total_loss");
        trace["total_loss_flagged"] = totalLoss;
        reasons.Add(new("NOT_TOTAL_LOSS", !totalLoss,
            !totalLoss
                ? "Not classified as total loss."
                : "Total loss classified — requires salvage/valuation review (WOK / RDW)."));

        // 9. Citation integrity (FR-11). A legal basis the corpus cannot resolve is
        // an unverifiable decision, so it blocks STP and goes to a handler.
        trace["citation_integrity"] = citationIntegrity;
        var citationsOk = citationIntegrity >= 1.0;
        reasons.Add(new("CITATIONS_RESOLVABLE", citationsOk,
            citationsOk
                ? "Every legal citation resolves to a retrieved corpus passage."
                : $"Citation integrity {citationIntegrity:P0} — at least one legal reference could not be "
                + "resolved against the retrieved corpus. Manual review.",
            "NFR-2 audit"));

        // 10. AI service health (NFR-6). A stage that fell back means the case file
        // is incomplete, so auto-approval is withheld — but the claim still gets a
        // decision and a handler, rather than an error page.
        trace["ai_services_healthy"] = aiServicesHealthy;
        reasons.Add(new("AI_SERVICES_HEALTHY", aiServicesHealthy,
            aiServicesHealthy
                ? "All AI stages completed."
                : "One or more AI stages degraded to fallback. Handler review before release."));

        // 11. Supporting evidence.
        var hasPhoto = documents.Any(d => d.DocType == "photo");
        var hasEstimate = documents.Any(d => d.DocType == "repair_estimate");
        reasons.Add(new("EVIDENCE_MINIMUM", hasPhoto && hasEstimate,
            hasPhoto && hasEstimate
                ? "At least one photo and one repair estimate on file."
                : "Missing photo and/or repair estimate. Adjuster to request."));

        var allPass = reasons.All(r => r.Ok);
        var hardFail = reasons.Any(r => !r.Ok && HardBlockers.Contains(r.Code));
        var outcome = allPass ? "auto_approved" : hardFail ? "manual" : "assisted";

        return new RulesResult(outcome, reasons, trace);
    }

    /// Additive fraud scoring. Score ≥ 0.30 blocks auto-approval via FRAUD_SIGNALS_LOW.
    /// Fraud indicators are investigation signals only — never an autonomous denial
    /// (Verbond van Verzekeraars / PIFI).
    ///
    /// ponytail: additive heuristic. Upgrade path — swap for a trained classifier
    /// once leakage / false-approval telemetry justifies it.
    public static FraudResult ComputeFraud(Claim claim, IReadOnlyList<Document> documents,
                                           int duplicatePhotoHits,
                                           IReadOnlyList<Signal>? photoSignals = null)
    {
        var signals = new List<Signal>();
        var score = 0.0;

        // 1. Recycled photo across prior claims — strongest single signal.
        if (duplicatePhotoHits > 0)
        {
            score += 0.6;
            signals.Add(new("PHOTO_RECYCLED", "high",
                $"Photo perceptual-hash matched {duplicatePhotoHits} document(s) in prior claim(s).", 0.6));
        }

        // 2. Delayed reporting.
        if (claim.LossDate is { } loss)
        {
            var days = DateOnly.FromDateTime(DateTime.UtcNow).DayNumber - loss.DayNumber;
            if (days > 180)
            {
                score += 0.2;
                signals.Add(new("REPORT_VERY_LATE", "medium",
                    $"Reported {days} days after loss (>180d).", 0.2));
            }
            else if (days > 60)
            {
                score += 0.1;
                signals.Add(new("REPORT_LATE", "low",
                    $"Reported {days} days after loss (>60d).", 0.1));
            }
        }

        // 3. No supporting evidence at all.
        if (documents.Count == 0)
        {
            score += 0.15;
            signals.Add(new("NO_EVIDENCE", "medium", "No documents or photos submitted.", 0.15));
        }

        // 4. Layered risk: injury plus third party.
        if (claim.Injuries && claim.ThirdPartyInvolved)
        {
            score += 0.1;
            signals.Add(new("INJURY_THIRD_PARTY_LAYERED", "medium",
                "Personal injury AND third-party liability reported — layered exposure.", 0.1));
        }

        // 5. Photo-level signals surfaced at upload (EXIF mismatch, no EXIF).
        //    Collapsed by code: ten photos with no EXIF is one anomaly seen ten
        //    times, not ten anomalies. Charging per photo made the score a function
        //    of how many pictures the claimant happened to send.
        foreach (var group in (photoSignals ?? []).GroupBy(p => p.Code))
        {
            var worst = group.OrderByDescending(p => SeverityWeight(p.Severity)).First();
            var weight = SeverityWeight(worst.Severity);
            var count = group.Count();
            score += weight;
            signals.Add(worst with
            {
                Weight = weight,
                Message = count == 1 ? worst.Message : $"{worst.Message} ({count} photos affected)",
            });
        }

        // 6. Very high estimate with low extraction confidence.
        var amount = claim.EstimatedAmountEur ?? 0;
        var conf = claim.ExtractionConfidence ?? 0;
        if (amount > 5000 && conf > 0 && conf < 0.6)
        {
            score += 0.1;
            signals.Add(new("AMOUNT_HIGH_CONF_LOW", "medium",
                $"Estimate €{amount:F0} with low extraction confidence ({conf:F2}).", 0.1));
        }

        return new FraudResult(Math.Min(score, 1.0), signals);
    }
}
